### PR TITLE
First approach to solve regression problem

### PR DESCRIPTION
**Pull request by kprokofi**
Pipeline debugged and working (at least with code point of view)

new data loader for COCO format
a model with two heads --> classification and regression
Module, class refactoring
config approach added
evaluation script added
metrics fixed

### CODE REVIEW

File: dataloaders/objectron_main.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+    def __getitem__(self, indx):
+        # get path to image from annotations
+        raw_keypoints = self.ann['annotations'][indx]['keypoints']
+        img_id = self.ann['annotations'][indx]['image_id']
+        category = self.ann['annotations'][indx]['category_id']
+        # get raw key points for bb from annotations
+        img_path = self.root_folder + '/' + (self.ann['images'][img_id]['file_name'])
+        # read image
+        image = cv.imread(img_path, flags=1)
```
Comment (sovrasov) [inline dataloaders/objectron_main.py]: Please use a corresponding constant from cv.* here instead of a bare integer

File: dataloaders/objectron_main.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            cv.imwrite('image_before_pipeline.jpg', imgh)
+        # given unnormalized keypoints crop object on image
+        cropped_keypoints, cropped_img = self.crop(image, unnormalized_keypoints)
+
+        # convert colors from BGR to RGB
+        cropped_img = cv.cvtColor(cropped_img, cv.COLOR_BGR2RGB)
+        # do augmentations with keypoints
+        if self.transform:
+            transformed = self.transform(image=cropped_img, keypoints=cropped_keypoints)
+            assert np.array(transformed['keypoints']).shape == np.random.rand(9,2).shape
```
Comment (sovrasov) [inline dataloaders/objectron_main.py]: shape is just a tuple, so I guess we don't need to generate a random array every time

File: main.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    # translate output to log file
+    log_name = 'train.log'
+    log_name += time.strftime('-%Y-%m-%d-%H-%M-%S')
+    sys.stdout = Logger(osp.join(cfg.output_dir, log_name))
+
+    # init main components
+    net = mobilenetv3_large(pretrained=True)
+    net.to(args.device)
+    if torch.cuda.is_available():
```
Comment (sovrasov) [inline main.py]: What if --device=cpu ?

File: scripts/evaluate.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        else:
+            assert isinstance(self.samples, list)
+            indexes = self.samples
+
+        for idx in tqdm(indexes):
+            # pull next data
+            prefetch_img, img, gt_kp, gt_cat = ds[idx]
+            # feed forward net and put it on cpu
+            pred_kp, pred_cat = self.model(torch.unsqueeze(img,0))
+            pred_kp = pred_kp.to('cpu')
```
Comment (sovrasov) [inline scripts/evaluate.py]: Let's mimic pytorch behavior: loss and similar functions don't force user to switch devices of the input tensors explicitly.

File: scripts/train_val.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            if self.debug and it == 20:
+                break
+
+        if self.save_chkpt:
+            save_snap(self.model, self.optimizer, epoch, self.log_path)
+
+        print(f"train: epoch: {epoch}, ADD: {ADD_meter.avg},"
+              f" SADD: {SADD_meter.avg}, loss: {losses.avg}, accuracy: {ACC_meter.avg}")
+
+    def val(self, epoch):
```
Comment (sovrasov) [inline scripts/train_val.py]: Does this code duplicates Tester.compute_metrics?

File: scripts/utils.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    cfg_dict = adict({
+        name: value
+        for name, value in mod.__dict__.items()
+        if not name.startswith('__')
+    })
+
+    return cfg_dict
+
+def collate(batch):
+    imgs = np.array([np.transpose(np.array(img), (2,0,1)).astype(np.float32)
```
Comment (sovrasov) [inline scripts/utils.py]: Seems like transpose() has already been done in `__getitem__`

File: dataloaders/objectron_main.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        transformed_bbox = self.normalize(transformed_image, transformed_bbox)
+        # "print" image after crop with keypoints if needed
+        if self.debug_mode:
+            imgh = transformed_image.copy()
+            b = np.zeros((9,3))
+            b[:,:2] = transformed_bbox
+            graphics.draw_annotation_on_image(imgh, b , [9])
+            cv.imwrite('after_preprocessing.jpg', cv.cvtColor(imgh, cv.COLOR_RGB2BGR))
+
+        transformed_image = np.transpose(transformed_image, (2,0,1)).astype(np.float32)
```
Comment (sovrasov) [inline dataloaders/objectron_main.py]: I'd propose to move all "to tensor" and related stuff to a albumentation-style transformation. Debugging mode could be also implemented in this way.

File: dataloaders/objectron_main.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+import torch
+import cv2 as cv
+import json
+import numpy as np
+from icecream import ic
+from numba import jit
+import albumentations as A
+import sys
+
+sys.path.insert(1, '/home/prokofiev/Objectron')
```
Comment (sovrasov) [inline dataloaders/objectron_main.py]: Remove the absolute path

File: scripts/train_val.py (hunk 5 above)
Comment (kprokofi) [reply scripts/train_val.py]: Deleted Val in Trainer.py. This method is available only in Evaluator now

**Comment by kprokofi** [comment]
@sovrasov, It seems to be done with all pending changes. Please, review one more time

File: setup.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+
+def find_version():
+    version_file = 'torchdet3d/version.py'
+    with open(version_file, 'r') as f:
+        exec(compile(f.read(), version_file, 'exec'))
+    return locals()['__version__']
+
+
+def numpy_include():
```
Comment (sovrasov) [inline setup.py]: Why do you need numpy_include files? I think it's required in case of compiling cython extensions, but we don't have any.

File: annotation_converters/objectron_helpers.py [hunk 10]
```diff
@@ -10,7 +10,7 @@
 from google.protobuf import text_format
 
 import sys
-module_path = os.path.abspath(os.path.join('3rdparty/Objectron'))
+module_path = os.path.abspath(os.path.join('./3rdparty/Objectron'))
```
Comment (sovrasov) [inline annotation_converters/objectron_helpers.py]: Seems like I've mistaken here: join with one argument is useless. Here should be something like
```
module_path = os.path.abspath(os.path.join(os.path.dirname(__file__), '../3rdparty/Objectron'))
```
to allow resolving path independently from the current working directory.

File: torchdet3d/__init__.py [hunk 11]
```diff
@@ -0,0 +1,14 @@
+from __future__ import absolute_import, print_function
+import sys
+import os
+
+module_path = os.path.abspath(os.path.join('./3rdparty/Objectron'))
```
Comment (sovrasov) [inline torchdet3d/__init__.py]: The same issue with this join as in the first comment. Even if we use `os.path.dirname(__file__)`, an issue with hardcoded Objectron location will remain, though it's not related to out usecase.

File: requirements.txt [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+numpy
+sklearn
+tqdm
+matplotlib
+tensorboard
+onnx
+glog
+torchvision
+torch
+gdown
```
Comment (sovrasov) [inline requirements.txt]: icecream

File: torchdet3d/evaluation/metrics.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
+import torch
+from icecream import ic
```
Comment (sovrasov) [inline torchdet3d/evaluation/metrics.py]: Useless import

File: torchdet3d/builders/loader_builder.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+def build_augmentations(cfg):
+    normalize = A.augmentations.transforms.Normalize (**cfg.data.normalization)
+
+    train_transform = A.Compose([
+                            A.Resize(*cfg.data.resize),
+                            A.HorizontalFlip(p=0.5),
+                            A.augmentations.transforms.ISONoise(color_shift=(0.15,0.35),
+                                                                intensity=(0.2, 0.5), p=0.2),
+                            normalize
```
Comment (sovrasov) [inline torchdet3d/builders/loader_builder.py]: Aren't ConvertColor and ToTensor  supposed to be here?

File: torchdet3d/dataloaders/objectron_main.py [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+        img_tensor, bbox, cat = next(iter_dt)
+        ic(mode)
+        ic(cat)
+        ic(img_tensor.shape)
+        ic(bbox.shape)
+        assert img_tensor.shape == (batch_size, 3, 290, 290)
+        assert bbox.shape == (batch_size, 9, 2)
+
+    root = 'data'
+    transform = A.Compose([ ConvertColor(),
```
Comment (sovrasov) [inline torchdet3d/dataloaders/objectron_main.py:145]: ConvertColor should be right before ToTensor. Early channel swapping cause misinterpreting colors by the other opencv functions you may apply after as augmentations.

Comment (kprokofi) [reply torchdet3d/dataloaders/objectron_main.py:145]: What about color transformation? Wouldn't it distort color spaces and do expected augmentation in the wrong way?

Comment (sovrasov) [reply torchdet3d/dataloaders/objectron_main.py:145]: This is exactly that happening now: all the opencv's color transforms assume BGR layout whereas you convert the image to RGB from the beginning.

Comment (kprokofi) [reply torchdet3d/dataloaders/objectron_main.py:145]: according to the Albumentation documentation and their examples, an augmentation pipeline should receive RGB format.
Please, refer to this [paragraph](https://albumentations.ai/docs/getting_started/image_augmentation/#step-3-read-images-from-the-disk)
So, I suggest keeping ConverColor first
